clients/lsp-csharp.el

```diff
@@ -359,6 +360,74 @@ using the `textDocument/references' request."
                                              ("o#/projectdiagnosticstatus" 'ignore))
                   :download-server-fn #'lsp-csharp--download-server))
 
+;;
+;; Alternative "csharp-ls" language server support
+;; see https://github.com/razzmatazz/csharp-language-server
+;;
+(lsp-defun lsp-csharp--csharp-ls-metadata-uri-handler (uri)
+  "Handle `csharp:/(metadata)' uri from csharp-ls server.
+
+'csharp/metadata' request is issued to retrieve metadata from the server.
+A cache file is created on project root dir that stores this metadata and filename
+is returned so lsp-mode can display this file."
+
+  (-when-let* ((metadata-req (lsp-make-csharp-ls-c-sharp-metadata
+                              :text-document (lsp-make-text-document-identifier :uri uri)))
+               (metadata (lsp-request "csharp/metadata" metadata-req))
+               ((&csharp-ls:CSharpMetadataResponse :project-name
+                                                   :assembly-name
+                                                   :symbol-name
+                                                   :source) metadata)
+               (filename (f-join ".cache"
+                                 "lsp-csharp"
+                                 "metadata"
+                                 "projects" project-name
+                                 "assemblies" assembly-name
+                                 (concat symbol-name ".cs")))
+               (file-location (expand-file-name filename (lsp--suggest-project-root)))
+               (metadata-file-location (concat file-location ".metadata-uri"))
+               (path (f-dirname file-location)))
+
+    (unless (find-buffer-visiting file-location)
+      (unless (file-directory-p path)
+        (make-directory path t))
+
+      (with-temp-file metadata-file-location
+        (insert uri))
+
+      (with-temp-file file-location
+        (insert source)))
+
+    file-location))
+
+(defun lsp-csharp--csharp-ls-before-file-open (_workspace)
+  "Set `lsp-buffer-uri' variable after C# file is open from *.metadata-uri file."
+
+  (let ((metadata-file-name (concat buffer-file-name ".metadata-uri")))
+    (setq-local lsp-buffer-uri
+                (when (file-exists-p metadata-file-name)
+                  (with-temp-buffer (insert-file-contents metadata-file-name)
+                                    (buffer-string))))))
+
+(lsp-register-client
+ (make-lsp-client :new-connection (lsp-stdio-connection '(lambda () "csharp-ls"))
+                  :priority -2
+                  :server-id 'csharp-ls
+                  :major-modes '(csharp-mode csharp-tree-sitter-mode)
+                  :before-file-open-fn #'lsp-csharp--csharp-ls-before-file-open
+                  :uri-handlers (ht ("csharp" #'lsp-csharp--csharp-ls-metadata-uri-handler))
+                  :download-server-fn (lambda (_client callback error-callback update?)
```

Comment:
@yyoncho does `update?` signal what I think it does (i.e. signals if the user wants to update the server?

because I could not find `M-x lsp-update-server` and python and F# LS have their own custom interactive commands.. so no so sure..

Reply:
It is triggered via `C-u M-x lsp-install-server`